controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -322,20 +327,22 @@ public void getDelegationToken(io.pravega.controller.stream.api.grpc.v1.Controll
                                    io.grpc.stub.StreamObserver<io.pravega.controller.stream.api.grpc.v1.Controller.DelegationToken> responseObserver)  {
         log.info("createStream called for stream {}/{}.", request.getScope(),
                 request.getStream());
-        authenticateExecuteAndProcessResults(v -> checkAuthorizationWithToken(request.getScope() + "/" +
+        authenticateExecuteAndProcessResults(() -> checkAuthorizationWithToken(request.getScope() + "/" +
                         request.getStream(), AuthHandler.Permissions.READ_UPDATE),
-                () -> CompletableFuture.completedFuture(Controller.DelegationToken
+                delegationToken -> CompletableFuture.completedFuture(Controller.DelegationToken
                         .newBuilder()
-                        .setDelegationToken(getCurrentDelegationToken())
+                        .setDelegationToken(delegationToken)
                         .build()),
                 responseObserver);
     }
 
     // Convert responses from CompletableFuture to gRPC's Observer pattern.
-    private static <T> void authenticateExecuteAndProcessResults(Predicate<Void> authenticator,
-                                                                 Supplier<CompletableFuture<T>> call, final StreamObserver<T> streamObserver) {
-        if (authenticator.test(null)) {
-            CompletableFuture<T> result = call.get();
+    private static <T> void authenticateExecuteAndProcessResults(Supplier<String> authenticator,
+                                                                 Function<String, CompletableFuture<T>> call, final StreamObserver<T> streamObserver) {
+        try {
+            String delegationToken;
```

Comment:
Is it resolving #2816 because we have a `delegationToken` per invocation of this method?

Reply:
Yes. This is correct. I think the updated description covers this. Do let me know if you know any further details.